[mlir][spirv] Add definition for OpEmitVertex and OpEndPrimitive

IgWod-IMG opened the pull request:
Hey,

This is hopefully the first patch in the series of patches adding some missing SPIR-V ops to MLIR over the next weeks/months, starting with something simple: `OpEmitVertex` and `OpEndPrimitive`. Since the ops have no input and outputs, and the only condition is "This instruction must only be used when only one stream is present.", which I don't think can be validate at the instruction level in isolation, I set `hasVerifier` to 0. I hope I didn't miss anything, but I'm more than happy to address any comments.

Also, I don't have commit rights, so someone else needs to merge it, once approved.

Many thanks,
Igor

IgWod committed:
[mlir][spirv] Add definition for OpEmitVertex and OpEndPrimitive

kuhar commented on mlir/include/mlir/Dialect/SPIRV/IR/SPIRVPrimitiveOps.td:
```diff
@@ -0,0 +1,87 @@
+    Capability<[SPIRV_C_Geometry]>
+  ];
+
+  let arguments = (ins);
+
+  let results = (outs);
+
+  let hasVerifier = 0;
+
+  let assemblyFormat = "attr-dict";
```
```suggestion
  let arguments = (ins);
  let results = (outs);
  let hasVerifier = 0;
  let assemblyFormat = "attr-dict";
```

kuhar commented on mlir/include/mlir/Dialect/SPIRV/IR/SPIRVPrimitiveOps.td:
```diff
@@ -0,0 +1,87 @@
+
+def SPIRV_EndPrimitiveOp : SPIRV_Op<"EndPrimitive", []> {
+  let summary = [{
+    Finish the current primitive and start a new one.  No vertex is emitted.
```
```suggestion
    Finish the current primitive and start a new one. No vertex is emitted.
```

kuhar commented on mlir/test/Target/SPIRV/primitive-ops.mlir:
```diff
@@ -0,0 +1,14 @@
+// RUN: mlir-translate -no-implicit-module -test-spirv-roundtrip %s | FileCheck %s
+
+spirv.module Logical GLSL450 requires #spirv.vce<v1.0, [Shader], []> {
+  spirv.func @emit_vertex() "None" {
+    // CHECK: spirv.EmitVertex
+    spirv.EmitVertex
```
Don't these require the Geometry execution model and capability?

kuhar reviewed (commented):
Is the plan to allow non-compute shaders?

IgWod-IMG commented:
Hey @kuhar,

Thank you for you feedback!

> Is the plan to allow non-compute shaders?

Yes and no, we use SPIR-V MLIR in the front-end of our research tool and we're interested in all sorts of shader, not only compute, so in that sense yes. However, we don't need intend to enable them in a runtime, e.g., provide further lowering and enabling it with `mlir-cpu-runner`, so in that sense no.

> Don't these require the Geometry execution model and capability?

Good point. I believe it should specify Geometry capability as well:

```mlir
spirv.module Logical GLSL450 requires #spirv.vce<v1.0, [Shader, Geometry], []> {
```

Do I need `ExecutionMode` as well? E.g.,

```mlir
  spirv.ExecutionMode @main "Triangles"
  spirv.ExecutionMode @main "OutputTriangleStrip"
  spirv.ExecutionMode @main "OutputVertices", 12
  spirv.ExecutionMode @main "Invocations", 1
```

In that case, would it make sense to merge both functions together, since having `EmitVertex` and `EndPrimitive` separately does not make much sense? I am just not clear on the level of detail this tests needs.

Once I have a clarity on that I will push an update (that also addresses formatting issues).

Many thanks,
Igor

kuhar commented:
> In that case, would it make sense to merge both functions together, since having `EmitVertex` and `EndPrimitive` separately does not make much sense? I am just not clear on the level of detail this tests needs.

What do you mean by 'merge both functions'?

> Do I need ExecutionMode as well? E.g.,

I'm not sure, can you write a compute shader that uses these ops (say, in glsl/hlsl) and make it pass validation with `spirv-val`?

Also note that we have tests for op availability: https://github.com/llvm/llvm-project/blob/main/mlir/test/Dialect/SPIRV/IR/availability.mlir .

IgWod-IMG commented:
> What do you mean by 'merge both functions'?

Sorry for the confusion. I meant merge both functions within the test. What I am proposing is to have test as below, instead of separating `EmitVertex` and `EndPrimitive` into two functions in the test:

```mlir
spirv.module Logical GLSL450 requires #spirv.vce<v1.0, [Geometry], []> {
  spirv.func @emit_verted_end_primitive() "None" {
    // CHECK: spirv.EmitVertex
    spirv.EmitVertex
    // CHECK: spirv.EndPrimitive
    spirv.EndPrimitive
    spirv.Return
  }
}
```

This is still not a completely correct shader, though. There needs to be a store to an output so there is something for `EmitVertex` to emit. And so are the execution modes needed to tell the hardware how and what new primitives are emitted.

> I'm not sure, can you write a compute shader that uses these ops (say, in glsl/hlsl) and make it pass validation with spirv-val?

So, you can't have a compute shader with those two ops. It must be geometry. I based my observation about the `ExecutionMode` on running a valid, although complex, shader through `mlir-translate`. But let's take a step back. Is the idea of the Target test to have a minimum viable shader that includes the new op and would correctly run? A simple example that I can probably further minimize would look, in SPIR-V, something like https://godbolt.org/z/6P9vK8YfY.

The direct translation of that example with `mlir-translate` is:

```mlir
spirv.module Logical GLSL450 requires #spirv.vce<v1.3, [Geometry], []> {
  spirv.GlobalVariable @spirv_var_3 : !spirv.ptr<!spirv.struct<(vector<4xf32> [BuiltIn=0], f32 [BuiltIn=1], !spirv.array<1 x f32> [BuiltIn=3])>, Output>
  spirv.func @main() "None" {
    %spirv_var_3_addr = spirv.mlir.addressof @spirv_var_3 : !spirv.ptr<!spirv.struct<(vector<4xf32> [BuiltIn=0], f32 [BuiltIn=1], !spirv.array<1 x f32> [BuiltIn=3])>, Output>
    %cst0_si32 = spirv.Constant 0 : si32
    %0 = spirv.AccessChain %spirv_var_3_addr[%cst0_si32] : !spirv.ptr<!spirv.struct<(vector<4xf32> [BuiltIn=0], f32 [BuiltIn=1], !spirv.array<1 x f32> [BuiltIn=3])>, Output>, si32 -> !spirv.ptr<vector<4xf32>, Output>
    %cst_vec_4xf32 = spirv.Constant dense<[-1.000000e-01, 0.000000e+00, 0.000000e+00, 0.000000e+00]> : vector<4xf32>
    spirv.Store "Output" %0, %cst_vec_4xf32 : vector<4xf32>
    spirv.EmitVertex
    %spirv_var_3_addr_0 = spirv.mlir.addressof @spirv_var_3 : !spirv.ptr<!spirv.struct<(vector<4xf32> [BuiltIn=0], f32 [BuiltIn=1], !spirv.array<1 x f32> [BuiltIn=3])>, Output>
    %cst0_si32_1 = spirv.Constant 0 : si32
    %1 = spirv.AccessChain %spirv_var_3_addr_0[%cst0_si32_1] : !spirv.ptr<!spirv.struct<(vector<4xf32> [BuiltIn=0], f32 [BuiltIn=1], !spirv.array<1 x f32> [BuiltIn=3])>, Output>, si32 -> !spirv.ptr<vector<4xf32>, Output>
    %cst_vec_4xf32_2 = spirv.Constant dense<[1.000000e-01, 0.000000e+00, 0.000000e+00, 0.000000e+00]> : vector<4xf32>
    spirv.Store "Output" %1, %cst_vec_4xf32_2 : vector<4xf32>
    spirv.EmitVertex
    spirv.EndPrimitive
    spirv.Return
  }
  spirv.EntryPoint "Geometry" @main, @spirv_var_3
  spirv.ExecutionMode @main "InputPoints"
  spirv.ExecutionMode @main "Invocations", 1
  spirv.ExecutionMode @main "OutputLineStrip"
  spirv.ExecutionMode @main "OutputVertices", 2
}
```

> Also note that we have tests for op availability: https://github.com/llvm/llvm-project/blob/main/mlir/test/Dialect/SPIRV/IR/availability.mlir .

I'll have a look at those tests.

kuhar commented:
> What I am proposing is to have test as below, instead of separating EmitVertex and EndPrimitive into two functions in the test:

Sounds good. Not every op has to have it's own test module -- we can group and test similar ops together.

>  But let's take a step back. Is the idea of the Target test to have a minimum viable shader that includes the new op and would correctly run?

It doesn't need to run and produce any results, but it should at the very least compile and validate.

IgWod-IMG commented:
Oh! That makes perfect sense now! Thank you for bearing with me. I got confused about what and how SPIR-V and MLIR gets verified, and this made me misunderstand your comments and what's the goal here.

Anyway, I'm going to update the test so it produces a valid SPIR-V (I think my merged example is almost correct - it may only need an `EntryPoint` for it to serialize into a valid SPIR-V), look into the availability test, clean up and push an updated commit in the next day or two.

Again, thanks a lot for the help, it was a valuable learning experience.

IgWod-IMG commented:
@kuhar I've just pushed an updated commit. Hopefully, it all should be good now. I addressed the formatting issues and added availability tests. For the Target test I created a minimal geometry shader that serializes into SPIR-V and passes validation with `spiv-val`. Let me know if there is anything else you would like me to address.

kuhar commented on mlir/test/Target/SPIRV/primitive-ops.mlir:
```diff
@@ -0,0 +1,17 @@
+// RUN: mlir-translate -no-implicit-module -test-spirv-roundtrip %s | FileCheck %s
```
```suggestion
// RUN: mlir-translate --no-implicit-module --test-spirv-roundtrip %s | FileCheck %s
```

kuhar commented on mlir/test/Dialect/SPIRV/IR/primitive-ops.mlir:
```diff
@@ -0,0 +1,21 @@
+// RUN: mlir-opt -split-input-file -verify-diagnostics %s | FileCheck %s
```
Right now this test doesn't use splits or diagnostics -- we can add those flags later when they become necessary

```suggestion
// RUN: mlir-opt %s | FileCheck %s
```

kuhar reviewed (approved):
LGTM % nits

IgWod-IMG commented:
Done! Please commit once ready, as I don't have committer rights. Thanks a lot!
